theano/common/wrapper.py

```diff
@@ -0,0 +1,448 @@
+
+Example of usage:
+    # Importation
+        >>> from theano.common import Wrapper, Wrap
+    # In a op you create:
+        >>> params_type = Wrapper(attr1=TensorType('int32', (False, False)), attr2=TensorType('float64', (True,False)))
+    # In the get_params() method of your op:
+        >>> return Wrap(attr1=numpyArray1, attr2=numpyArray2)
+    # In perform() implementation (with params named `param`):
+        >>> print(param.attr1)
+        >>> print(param.attr2)
+    # In c_code() implementation (with `param = sub['params']`):
+        ```
+        PyArrayObject* attr1 = param.attr1;
+        PyArrayObject* attr2 = param.attr2;
+        /* Just use attr1 and attr2, you won't need to free them or whatever else. */
+        ```
+
+See theano/common/tests/test_wrapper.py for a complete working example.
+
+"""
+
+from __future__ import absolute_import, print_function, division
+import re
+import hashlib
+import numpy
+from theano.gof.utils import MethodNotDefined
+from theano.gof import Type
+from theano.gof.cmodule import GCC_compiler as compiler
+
+# NB: Maybe we should check if an attribute name is a C/C++ keyword, and raise an error if so.
+# These are some lists of C/C++ keywords:
+# http://fr.cppreference.com/w/cpp/keyword
+# http://fr.cppreference.com/w/c/keyword
+
+
+class Wrap(object):
+    """
+    Convenient class to wrap many Python objects into one.
+
+    Example:
+        >>> w = Wrap(attr1=var1, attr2=var2, attri=vari)
+        >>> print(w.attr1, w.attr2, w.attri)
+        >>> d = dict(a=1, b=2, c='test')
+        >>> w2 = Wrap(**d)
+        >>> print(w2.a, w2.b, w2.c)
+
+    """
+
+    def __init__(self, **kwargs):
+        if len(kwargs) == 0:
+            raise TypeError('Wrap: cannot wrap empty data.')
+        super(Wrap, self).__setattr__('data', kwargs)
+
+    def __repr__(self):
+        return 'Wrap(%s)' % ', '.join([('%s:%s' % (k, self.data[k])) for k in sorted(self.data.keys())])
+
+    def __getattr__(self, key):
+        if key not in self.data:
+            raise AttributeError('Wrap: attribute "%s" does not exist.' % key)
+        return self.data[key]
+
+    def __setattr__(self, key, value):
+        if key not in self.data:
+            raise AttributeError('Wrap: attribute "%s" does not exist.' % key)
+        self.data[key] = value
+
+    def __hash__(self):
+        keys = sorted(self.data.keys())
+        types = []
+        attributes = []
+        for k in keys:
+            types += (type(self.data[k]),)
+            if isinstance(self.data[k], numpy.ndarray):
+                if len(self.data[k].shape) == 0:
+                    attributes += (numpy.asscalar(self.data[k]),)
+                else:
+                    attributes += tuple(self.data[k])
+            else:
+                try:
+                    iter(self.data[k])
+                except TypeError:
+                    attributes += (self.data[k],)
+                else:
+                    attributes += tuple(self.data[k])
```

Comment:
I assume that the values of op params can be lists or tuples, that could be converted later to ndarrays when passed to `theano_type.filter(data)` when `strict==False`, for example. So I try to support any iterable object here.

But indeed, I don't need to cast to tuple if the object is already iterable. I can directly write `attributes += self.data[x]`.